sql/1715118159.sql

```diff
@@ -0,0 +1,112 @@
+-- +migrate Up
+-- +migrate StatementBegin
+create or replace function convoy.take_token(_key text, _rate integer, _bucket_size integer) returns boolean
+    language plpgsql
+as
+$$
+DECLARE
+    next_min timestamptz;
+    can_take BOOLEAN;
+    row RECORD;
+BEGIN
+    SELECT expires_at, tokens FROM convoy.token_bucket WHERE key = _key FOR UPDATE SKIP LOCKED LIMIT 1 INTO row;
+    if row is null then
+        return false;
+    end if;
+
+    next_min := current_timestamp + make_interval(secs := _bucket_size);
+
+    IF current_timestamp < row.expires_at AND row.tokens = _rate THEN
+        RETURN FALSE;
+    END IF;
+
+    -- Update existing record
+    UPDATE convoy.token_bucket
+    SET tokens =
+            CASE WHEN current_timestamp > expires_at
+                     THEN 1
+                 ELSE CASE WHEN tokens < _rate
+                               THEN tokens + 1
+                           ELSE tokens END
+                END,
+        expires_at =
+            CASE WHEN current_timestamp > expires_at
+                     THEN next_min
+                 ELSE CASE WHEN tokens < _rate
+                               THEN next_min
+                           ELSE expires_at
+                     END
+                END,
+        rate = COALESCE(_rate, rate),
+        updated_at = DEFAULT
+    WHERE key = _key
+    RETURNING TRUE INTO can_take;
+
+    -- Insert if no record found
+    IF NOT FOUND THEN
```

Comment:
@jirevwe, what if we bootstrap the row for each endpoint at app startup? We won't need this anymore.

Reply:
The table is unlogged. It will be wiped if the VM it's running on is restarted or recycled. So it's best to create the record lazily.